### PR TITLE
Reduce update

### PR DESCRIPTION
**Pull request by jagrit06**
## Proposed changes

* Split up reduction kernels to reduce compile time
* Add specializations for small reduction axes in row reductions

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

**Comment by jagrit06** [comment]
Here are some numbers:

Before: 

(base) jdigani@mlx-m2-macstudio-2 mlx % python benchmarks/python/comparative/bench_mlx.py sum_axis --size 32768x4 --axis 1 
0.4660329818725586
(base) jdigani@mlx-m2-macstudio-2 mlx % python benchmarks/python/comparative/bench_mlx.py sum_axis --size 32768x16 --axis 1 
0.47240376472473145
(base) jdigani@mlx-m2-macstudio-2 mlx % python benchmarks/python/comparative/bench_mlx.py sum_axis --size 32768x32 --axis 1 
0.4732379913330078
(base) jdigani@mlx-m2-macstudio-2 mlx % python benchmarks/python/comparative/bench_mlx.py sum_axis --size 32768x64 --axis 1 
0.4062469005584717
(base) jdigani@mlx-m2-macstudio-2 mlx % python benchmarks/python/comparative/bench_mlx.py sum_axis --size 32768x128 --axis 1 
0.4185929298400879
(base) jdigani@mlx-m2-macstudio-2 mlx % python benchmarks/python/comparative/bench_mlx.py sum_axis --size 32768x256 --axis 1 
0.529871940612793

After: 

(base) jdigani@mlx-m2-macstudio-2 mlx % python benchmarks/python/comparative/bench_mlx.py sum_axis --size 32768x4 --axis 1  
0.19516921043395996
(base) jdigani@mlx-m2-macstudio-2 mlx % python benchmarks/python/comparative/bench_mlx.py sum_axis --size 32768x16 --axis 1  
0.19518685340881348
(base) jdigani@mlx-m2-macstudio-2 mlx % python benchmarks/python/comparative/bench_mlx.py sum_axis --size 32768x32 --axis 1  
0.18773913383483887
(base) jdigani@mlx-m2-macstudio-2 mlx % python benchmarks/python/comparative/bench_mlx.py sum_axis --size 32768x64 --axis 1  
0.1961989402770996
(base) jdigani@mlx-m2-macstudio-2 mlx % python benchmarks/python/comparative/bench_mlx.py sum_axis --size 32768x128 --axis 1 
0.2138679027557373
(base) jdigani@mlx-m2-macstudio-2 mlx % python benchmarks/python/comparative/bench_mlx.py sum_axis --size 32768x256 --axis 1 
0.27953100204467773

Reference: 

(base) jdigani@mlx-m2-macstudio-2 mlx % python benchmarks/python/comparative/bench_torch.py sum_axis --size 32768x4 --axis 1 
0.20731186866760254
(base) jdigani@mlx-m2-macstudio-2 mlx % python benchmarks/python/comparative/bench_torch.py sum_axis --size 32768x16 --axis 1 
0.21143603324890137
(base) jdigani@mlx-m2-macstudio-2 mlx % python benchmarks/python/comparative/bench_torch.py sum_axis --size 32768x32 --axis 1 
0.21271109580993652
(base) jdigani@mlx-m2-macstudio-2 mlx % python benchmarks/python/comparative/bench_torch.py sum_axis --size 32768x64 --axis 1 
0.3055598735809326
(base) jdigani@mlx-m2-macstudio-2 mlx % python benchmarks/python/comparative/bench_torch.py sum_axis --size 32768x128 --axis 1 
0.5393259525299072
(base) jdigani@mlx-m2-macstudio-2 mlx % python benchmarks/python/comparative/bench_torch.py sum_axis --size 32768x256 --axis 1 
0.5700762271881104

**Comment by angeloskath** [comment]
The refactoring is so nice and the speedup is fantastic! But I think we should address small reductions in general not just row reductions without out-of-row reduction. For instance the following can't use these kernels

```
python benchmarks/python/comparative/bench_mlx.py sum_axis --size 2x16384x128 --axis 0,2
python benchmarks/python/comparative/bench_mlx.py sum_axis --size 16x16384x16 --axis 0,2  # <--- particularly problematic
python benchmarks/python/comparative/bench_mlx.py sum_axis --size 8x32x2x32x2x128 --axis 2,4 # <--- of particular interest as it is most of our pooling ops
```

**Comment by jagrit06** [comment]
@angeloskath Yes! I'll add options to handle those as well

**Comment by jagrit06** [comment]
@angeloskath I think I will need a bit of an overhaul here to generically handle the non-row reductions in cases like 
`python benchmarks/python/comparative/bench_mlx.py sum_axis --size 16x16384x16 --axis 0,2`

For the moment we have gone from 

to 
```
(base) jdigani@mlx-m2-macstudio-2 mlx % python benchmarks/python/comparative/bench_mlx.py sum_axis --size 16x16384x16 --axis 0,2
0.44909214973449707
```

but PyTorch remains at 
```
(base) jdigani@mlx-m2-macstudio-2 mlx % python benchmarks/python/comparative/bench_torch.py sum_axis --size 16x16384x16 --axis 0,2
0.3736867904663086
```

This number isn't hard to beat, we just need to be able to skip the elem_to_loc since there is only one strided dimension. I could create a stop gap here but then the issue lands on how to speed up the higher dimensions too - I have some ideas but they require more specializations 

I think it might be best to merge this in and do those in a new PR so I can give a clean way of doing that some thought

File: mlx/backend/metal/kernels/reduction/ops.h [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// Copyright © 2023 Apple Inc.
+// Copyright © 2024 Apple Inc.
```
Comment (angeloskath) [inline mlx/backend/metal/kernels/reduction/ops.h:1]: The mother of all nitpicks but this should probably be 2023-2024 vs just 2024.

**Review by angeloskath — APPROVED** [review]
This looks absolutely fantastic.

I am not sure we need an overhaul but for sure we need to support the same in col reduce and I can't wait for it in a next PR.

**Comment by awni** [comment]
Better late than never: but really nice!!!

**Pull request merged**